### DISCUSSION TITLE
Joint position actuator question

### DISCUSSION THREAD

**Discussion by btaba** [google-deepmind/mujoco#754] [General]
Hi! I'm trying to understand the equation for qfrc_actuator for hinge/slide joints for a position joint actuator. I couldn't find a similar question, apologies if this was already asked somewhere.

Trying to write out the equation from the code:

gain is:
https://github.com/deepmind/mujoco/blob/fd1070cf039a586e0836d43baf2ff5bf4bd887b4/src/engine/engine_forward.c#L210

bias is:
https://github.com/deepmind/mujoco/blob/fd1070cf039a586e0836d43baf2ff5bf4bd887b4/src/engine/engine_forward.c#L251

actuator_length is:
https://github.com/deepmind/mujoco/blob/fd1070cf039a586e0836d43baf2ff5bf4bd887b4/src/engine/engine_core_smooth.c#L651

and $force = gain*ctrl + bias$ from [here](https://github.com/deepmind/mujoco/blob/fd1070cf039a586e0836d43baf2ff5bf4bd887b4/src/engine/engine_forward.c#L235) and [here](https://github.com/deepmind/mujoco/blob/fd1070cf039a586e0836d43baf2ff5bf4bd887b4/src/engine/engine_forward.c#L270).

Now for a position actuator, `gainprm = [kp, 0, 0]` and `biasprm = [0, -kp, 0]` from [here](https://mujoco.readthedocs.io/en/latest/XMLreference.html?highlight=gainprm#actuator-j-position).

plugging these all in we get:
$force = k_p * ctrl  - k_p * q * gear$

To get qfrc_actuator 
moment is:
https://github.com/deepmind/mujoco/blob/fd1070cf039a586e0836d43baf2ff5bf4bd887b4/src/engine/engine_core_smooth.c#L652

qfrc_actuator is:
https://github.com/deepmind/mujoco/blob/fd1070cf039a586e0836d43baf2ff5bf4bd887b4/src/engine/engine_forward.c#L300

so $qfrc = gear * force = gear * k_p * ctrl - gear^2 * k_p * q$

What's the intuition behind the $gear^2$ term (unless I made some mistake)? I would expect the equation to look something like $gear * k_p * (ctrl - q)$

Thanks for taking a look!!!

**Comment by Balint-H** [comment]
It's to convert between `ctrl`'s scale and the position's scale. `ctrl` defines the target in actuator length space, which is different from the joint position by a factor of `gear`. So your final expression could be: $gear * k_p * (ctrl-gear*q)$. This is same as having the $gear^2$ term for q.

**Reply by yuvaltassa** [reply]
This is a bit like the gear^2 term of [reflected inertia](https://www.sciencedirect.com/topics/engineering/referred-inertia).

The intuition is that you feel the effects of the gear **twice**, once on the way out, in the scaling of the angle and once again on the way in, when applying the force.

**Reply by btaba** [reply]
Thanks for the insight here!

**Reply by btaba** [reply]
(found this link to help understand reflected inertia https://physics.stackexchange.com/questions/73531/is-squared-motor-gearbox-ratio-proportional-to-inertia-ratio)

To help clarify a bit:

given $\tau_{in} = {gear} * \tau_{out}$ and $q_{out} = {gear} * q_{in}$

you are setting $\tau_{out} = {ctrl} - q_{out}$, which gives us $\tau_{in} = {gear} * ( {ctrl} - {gear} * q_{in} )$
